test/astarte/client/appengine/devices_test.exs

```diff
@@ -0,0 +1,476 @@
+        )
+      end)
+
+      Devices.get_datastream_data(client, @device_id, @sensors_stream_interface,
+        query: query_params
+      )
+    end
+
+    test "makes a request to expected url using expected method and passing query parameters with mapping path",
+         %{
+           client: client
+         } do
+      query_params = [limit: 1]
+
+      Tesla.Mock.mock(fn %{method: method, url: url, query: query} ->
+        assert method == :get
+
+        assert url ==
+                 build_interface_url(@device_id, @sensors_stream_interface,
+                   path: @sensor_path,
+                   query: query_params
+                 )
+
+        assert query_params == query
+
+        Tesla.Mock.json(
+          %{"data" => @sensor_stream_last_data},
+          status: 200
+        )
+      end)
+
+      Devices.get_datastream_data(client, @device_id, @sensors_stream_interface,
+        path: @sensor_path,
+        query: query_params
+      )
+    end
+
+    test "returns data from successful response", %{client: client} do
+      Tesla.Mock.mock(fn _ ->
+        Tesla.Mock.json(
+          %{"data" => @sensor_stream_data},
+          status: 200
+        )
+      end)
+
+      assert {:ok, body} =
+               Devices.get_datastream_data(client, @device_id, @sensors_stream_interface,
+                 path: @sensor_path
+               )
+
+      assert body["data"] == @sensor_stream_data
+    end
+
+    test "returns APIError on error", %{client: client} do
+      error_data = %{"errors" => %{"detail" => "Endpoint not found"}}
+      error_status = 400
+
+      Tesla.Mock.mock(fn _ ->
+        Tesla.Mock.json(error_data, status: error_status)
+      end)
+
+      assert {:error, %APIError{response: error_data, status: error_status}} ==
+               Devices.get_datastream_data(
+                 client,
+                 @device_id,
+                 @sensors_stream_interface,
+                 path: "/unexistingEndpoint"
+               )
+    end
+  end
+
+  describe "send_datastream/5" do
+    test "makes a request to expected url using expected method", %{client: client} do
+      data = %{
+        "value" => 20.3,
+        "timestamp" => "2022-02-01T12:00:03.837Z"
+      }
+
+      Tesla.Mock.mock(fn %{method: method, url: url, body: body} ->
+        assert method == :post
+        assert url == build_interface_url(@device_id, @sensors_interface, path: @sensor_path)
+
+        Tesla.Mock.json(
+          body,
+          status: 200
+        )
+      end)
+
+      Devices.send_datastream(client, @device_id, @sensors_interface, @sensor_path, data)
+    end
+  end
+
+  defp build_device_url(device_id) do
+    Path.join([@base_url, "appengine", "v1", @realm_name, "devices", device_id])
+  end
+
+  defp build_interface_url(device_id, interface, opts \\ []) do
+    path = Keyword.get(opts, :path, "")
+
+    Path.join([
+      build_device_url(device_id),
+      "interfaces",
+      interface
+    ]) <> path
```

Comment:
Why don't you pass `path` to `Path.join` too?

Reply:
We have convention that paths should have leading slash [PR](https://github.com/astarte-platform/astarte-client-elixir/pull/8). 
If we add `path` into `Path.join` we will have same output for
`build_interface_url(device_id, interface, path: "path")` and
`build_interface_url(device_id, interface, path: "/path")`.

Even if this function is used only for testing purpose I prefer it to have same behavior.

Reply:
Legit